### DISCUSSION TITLE
async chunk writing

### DISCUSSION THREAD

**Discussion by kongdd** [JuliaIO/Zarr.jl#125] [Q&A] [answered]
Just wondering whether Zarr.jl supports parallel writing? 
If yes, could you give one small example?

**Comment by meggart** [comment, marked as answer]
I am not really sure what the context is. It is completely fine if multiple processes write simultaneously into the same Zarr dataset, but it is up to the user/library to make sure that different processes/threads never write into the same chunk. So the following is fine:
````julia
using Zarr
a = zzeros(Float64,10,10,chunks=(10,1),path=tempname())
@sync for i=1:10
  @async a[:,i] = rand(10)
end
a[:,:]
````
However, when using multiple threads there may be problems coming from the different compression/decompression libraries, which are not thread-safe, so be careful when using compression. 

Otherwise when writing across several chunks in a single call to `setindex!`, it will depend on the backend if chunks are written sequentially or simulaneously. There is a storage [backend trait](https://github.com/JuliaIO/Zarr.jl/blob/74da9854b1a0bef648c7f3c3191cd3e6730a25f4/src/Storage/Storage.jl#L98-L107) which is set to `SequentialRead` by default, but several backends like HTTP or S3 do concurrent reads and writes by default, so if you do 

````julia
a[:,:] = rand(10,10)
````

it will depend on the backend if you do concurrent writes, but compression will always happen sequentially because of the problems with the underlying C libraries mentioned above.

**Comment by bjarthur** [comment]
> However, when using multiple threads there may be problems coming from the different compression/decompression libraries, which are not thread-safe, so be careful when using compression.

perhaps it's time to add [Blosc2](https://github.com/JuliaIO/Blosc2.jl) support to Zarr?  i believe that is thread safe, but could be wrong.
